doc/dev/ragged_iter_domain_design_doc.md

```diff
@@ -0,0 +1,914 @@
+2. Offsets can be derived from extents via cumulative sum
+3. Storing extents avoids redundancy (offsets contain cumulative information)
+
+**Offset Computation**: The `offsets()` method computes offsets on-demand as the cumulative sum of extents: `offsets[i] = sum(extents[0..i-1])`. This enables efficient indexing into contiguous storage:
+- Component 0 starts at offset 0
+- Component 1 starts at offset = extent[0]
+- Component 2 starts at offset = extent[0] + extent[1]
+- And so on...
+
+For a ragged dimension with extents [3, 5, 2], the computed offsets are [0, 3, 8, 10].
+
+**API Note**: While the internal representation uses extents, the public API continues to work with offsets for user convenience and compatibility with existing offset-based patterns (e.g., PyTorch nested tensors use offsets). The conversion between extents and offsets is handled transparently.
+
+#### Extent Semantics
+
+**extent()**: Overrides `IterDomain::extent()` to return the sum of all nested domain extents (total extent). This represents the total storage size needed for the ragged dimension when data is stored contiguously.
+
+### 6.3 Property Uniformity
+
+Certain IterDomain properties are allowed to be non-uniform across all nested domains, such as extents. In the initial version, properties that have no reason to be non-uniform are set to be uniform to simplify the overall design and implementation.
+
+| Property | Uniformity | Rationale |
+|----------|------------|-----------|
+| **extent** | **VARIABLE** | Core ragged characteristic - each component has different length |
+| **ParallelType** | **UNIFORM** | GPU execution model requires consistent thread mapping |
+| **IterType** | **UNIFORM** | All components perform same operation (iteration/reduction) |
+| **start** | **UNIFORM (=0)** | Simplifies offset computation; all components start at 0 |
+| **is_rfactor_domain** | **UNIFORM** | Reduction transformation applies uniformly |
+
+The following properties are out of scope of this initial buildout:
+- `is_padded_dimension`
+- `is_clustered_dimension`
+- `padded_to_size`
+
+The constructor validates that:
+1. Uniform properties (ParallelType, IterType, start, is_rfactor_domain) are consistent across all nested domains
+2. Out-of-scope properties are not set (must be false/nullptr for all nested domains)
+
+If any validation fails, an error is thrown.
+
+### 6.4 Key Operations
+
+#### Partition Operation (IterDomain-level)
+
+The `partition` operation is the fundamental primitive for creating ragged dimensions. It splits a regular IterDomain into a batch IterDomain and a RaggedIterDomain based on variable-length segments defined by offsets.
+
+**TensorView API:**
+```cpp
+// TensorView API (csrc/ir/interface_nodes.h)
+class TensorView : public Val {
+  // Partition dimension 'dim' using the provided offsets
+  // Returns new TensorView with partitioned dimension replaced by (batch_id, ragged_id)
+  TensorView* partition(int dim, TensorView* offsets);
```

Comment:
Why is `partition` a **member method** of TensorView instead of `TensorView* partition(TensorView* in, int dim, TensorView* offsets)` as other TensorView operations?

Reply:
Yes, it's meant to be an iter domain op, much like `split` and ` merge`. To me, it's different enough to have a separate name than `split`. 

> This is more convenient when we have a series of pointwise operations applied on a nested TensorView -- we don't need each TensorView to have a partition from root to logical. 

I think that'd be the case no matter if it's `partition` or `split`. What am I missing?

Reply:
I was trying to say `partition` or `split`, the IterDomain op, should **not** appear in **every** nested TenesorView. 

```
       s
      / \ split
in: [e, j1]
     |
     | cos
     v
out: [e, j1]  # no `s` in the root domain
```

This way, we can do the following merging reshape. Note `j2`'s definition is a merge not a partition/split, despite that `j2`'s **extent** is an extent TensorView. 
```
     s
    / \ split
  [e, j1]
     |
     | cos
     v
  [e, j1]
     |
     | reshape
     v
  e,  j1
   \ /
   [j2]
```

Reply:
Great -- I think we are on the same page. 

IIUC, Partition, an IterDomain op, takes **one** IterDomain and produces **two** IterDomains. It does **not** take a TensorView -- that's embedded (as extent) in the second output which is a RaggedIterDomain. This is why we can reuse Split but I'm fine with a different op. Btw, do you need a different op for merging a regular IterDomain and a RaggedIterDomain and call it maybe `combine`?

Reply:
Yeah, that feels a little inconsistent -- a new name for production but overloading the existing name for the reverse operation. My feeling is that `split` and `partition` have different input parameters: a scalar factor for `split` and an offset tensor for `partition`, but `merge` and `combine` would have the same set of inputs, i.e., two `IterDomain` pointers.

For the sake of consistency, I may introduce `combine` instead of extending `merge`.

Reply:
Could be, I may change my plan. That said, I'd start with its own class since it would work more naturally with `OptOutDispatcher` and I'd expect we would usually do something different for partitioning than splitting.